python/ray/tune/impl/tuner_internal.py

```diff
@@ -521,20 +525,30 @@ def _process_scaling_config(self) -> None:
     @classmethod
     def setup_create_experiment_checkpoint_dir(
         cls, trainable: TrainableType, run_config: Optional[RunConfig]
-    ) -> str:
-        """Sets up experiment checkpoint dir before actually running the experiment."""
-        path = Experiment.get_experiment_checkpoint_dir(
-            trainable,
-            # TODO(justinvyu): This is a result of the old behavior of parsing
-            # local vs. remote storage paths and should be reworked in a follow-up.
-            _get_defaults_results_dir()
-            if _use_storage_context()
-            else run_config.storage_path,
-            run_config.name,
-        )
-        if not os.path.exists(path):
-            os.makedirs(path, exist_ok=True)
-        return path
+    ) -> Tuple[str, str]:
+        """Sets up and creates the local experiment checkpoint dir.
+        This is so that the `tuner.pkl` file gets stored in the same directory
+        and gets synced with other experiment results.
+
+        Returns:
+            Tuple: (experiment_path, experiment_dir_name)
+        """
+        if _use_storage_context():
```

Comment:
Could we pass storage context here so the path resolution logic is in storage.py?

Reply:
Yeah, I was thinking about this. This logic is a bit weird because:
* `TunerInternal` is the outer shell that relies on `tune.run` for the actual implementation.
* `tune.run` handles the experiment dir name generation (if None, set to the trainable name, add a timestamp, etc.)
* `tune.run` is also where the storage ctx gets created, so we don't have access to it at the Tuner level.
* But, `TunerInternal` needs to save some file in the experiment dir, so it needs to determine the experiment dir ahead of time.

I was thinking we could create a storage context at the Tuner level and pass it into `tune.run`, but that's also confusing to add this internal argument to `tune.run`, then needing to deal with the case where a user calls tune.run directly vs. through the Tuner.

Sidenote: What is your opinion on making `tune.run` private?

Reply:
Hmm. It sounds like we should do some refactoring and pull out a private common run implementation, or something like this.

But it could probably wait for a future PR / 2.8.